### PR TITLE
[api][SIM-#235] Logger and global error handler

### PR DESCRIPTION
**Pull request by fwpushan**
---------------------------------------------------
1. Create LoggerService and Module
2. Create Decorator and interface to inject logger
4. Create Global error filter
5. Update classes with proper logger
6. Lint fixes

### CODE REVIEW

File: sources/packages/api/src/main.ts [hunk 1]
```diff
@@ -1,20 +1,41 @@
...
+  // Getting application port
   const port = process.env.PORT || 3000;
 
+  // Enabling cors for non prod env
   if (process.env.NODE_ENV !== "production") {
     app.enableCors();
   }
 
+  // Setting express middleware for req
+  app.use(LoggerService.apiLogger);
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/main.ts:37]: Is the idea to log all requests even on PROD env?
Right now we have only conole.log as the final log, but depending on how would be the final log storage maybe we do not want to log every request every time.
Is it a placeholder to be expanded in the future or even be enabled/disabled upon configuration?

File: sources/packages/api/src/database/database.service.spec.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+  it("should be defined", () => {
+    expect(service).toBeDefined();
+  });
+
+  it("should create service", async () => {
+    const connection = getConnection();
+    const svc = new DatabaseService(connection);
+    expect(svc).toBeDefined();
+    expect(svc.logger()).toBeDefined();
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/database/database.service.spec.ts:27]: Is it possible, using the current design, to replace the current logger implementation with a mocked object?
In the future, if our logger is saving to a file or a database, and I want to execute a unit test is some service that uses the logger, is it possible to have a mocked logger to test this service?

Comment (fwpushan) [reply sources/packages/api/src/database/database.service.spec.ts:27]: It is partial test of the concepts, but my idea is to mock it through LoggerService ....

Comment (fwpushan) [reply sources/packages/api/src/database/database.service.spec.ts:27]: Currently Nest logger is automatically disabled in test env

File: sources/packages/api/src/main.ts (hunk 1 above)
Comment (fwpushan) [reply sources/packages/api/src/main.ts:37]: YES, we need discussion on this topic. And I'm totally agree, logging each request could be overhead if persistent log storage introduce....

File: sources/packages/api/src/common/loggable.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import { LoggerService } from "../logger/logger.service";
+
+export interface Loggable {
+  logger(): LoggerService;
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/common/loggable.ts:4]: Is it possible to define a property instead of a method for the interface?

**Review by 2011sandeepraj — CHANGES_REQUESTED** [review]
As discussed, 
Base out classes we can. 
Inject LoggerService instead of having classes implement Loggable

File: sources/packages/api/src/common/loggable.ts (hunk 3 above)
Comment (fwpushan) [reply sources/packages/api/src/common/loggable.ts]: DONE

File: sources/packages/api/src/database/data.model.service.ts [hunk 4]
```diff
@@ -1,8 +1,10 @@
+import { LoggerService } from "../logger/logger.service";
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/database/data.model.service.ts:1]: Can you remove it, please?

File: sources/packages/api/src/services/auth/keycloak/keycloak.service.ts [hunk 5]
```diff
@@ -49,7 +54,7 @@ export class KeycloakService {
       return response.data as OpenIdConfig;
     } catch (ex) {
       // TODO: Add a logger.
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/auth/keycloak/keycloak.service.ts:56]: I believe that we can remove the // TODO: Add a logger.

**Review by andrewsignori-aot — APPROVED** [review]
👍

**Pull request merged**